### PR TITLE
Config: make writing to disk as atomic as possible

### PR DESCRIPTION
**Pull request by sphuber**
Potentially addresses #4604 

The original implementation of the `Config.store` method, that writes
the configuration from memory to disk, was using `shutil.copy` which is
not an atomic operation. It is therefore possible that if the copy
operation is interrupted the configuration is not written entirely and
is left either empty or in a corrupted state.

Making this an atomic operation in 100% of the cases is not trivial and
even on the application level there are no guarantees as even on the
file system level operations that should be atomic can be interrupted
under very rare circumstances.

However, we can make a significant improvement by replacing the use of
`shutil.copy` with `os.rename`, which because it is a file move
operation should be atomic and therefore less prone to being
interrupted. The new algorithm for storing the config is as follows:

 1. Create temporary file in same directory as target file
 2. Write content of in memory config to the temporary file and flush
 3. Use `os.rename` to move the temporary file to target destination

This should prevent most common cases that could corrupt the existing
configuration file.

Finally, in `Config.from_method` now only `FileNotFoundError` is caught
when trying to read the config file from disk. Originally both `IOError`
and `OSError` were being caught, which can also be raised if an existing
file could not be read. Simply overwriting in this case with a default
config is not desirable and so in this case we simply let the exception
bubble up.

### CODE REVIEW

File: aiida/manage/configuration/config.py [hunk 1]
```diff
@@ -378,38 +378,52 @@ def store(self):
...
-            with open(self.filepath, 'wb') as handle:
-                self._write(handle)
+            self._atomic_write()
             return self
 
         # Otherwise, we write the content to a temporary file and compare its md5 checksum with the current config on
         # disk. When the checksums differ, we first create a backup and only then overwrite the existing file.
         with tempfile.NamedTemporaryFile() as handle:
-            self._write(handle)
+            json.dump(self.dictionary, handle, indent=DEFAULT_CONFIG_INDENT_SIZE)
```
Comment (ltalirz) [inline aiida/manage/configuration/config.py:392]: is there a reason not to use `self._write` here?

File: tests/manage/configuration/test_config.py [hunk 2]
```diff
@@ -185,20 +184,32 @@ def compare_config_in_memory_and_on_disk(self, config, filepath):
...
+        """Test that ``Config.from_file`` does not overwrite if the content was not migrated."""
+        from time import sleep
+
+        # Construct the ``Config`` instance and write it to disk
+        Config(self.config_filepath, self.config_dictionary).store()
+
+        timestamp = os.path.getmtime(self.config_filepath)
+        Config.from_file(self.config_filepath)
+
+        # Sleep a second, because for some operating systems the time resolution is of the order of a second
```
Comment (ltalirz) [inline tests/manage/configuration/test_config.py:208]: Wow, which OS has 1s time resolution?
It's interesting that this can work with processors that have clock cycles below ns ;-)

As far as I'm aware, tests should run "in the blink of an eye", i.e. 1s is already too long.. 0.1s would already be better..

**Review by ltalirz — COMMENTED** [review]
thanks @sphuber , just some small comments

File: aiida/manage/configuration/config.py (hunk 1 above)
Comment (sphuber) [reply aiida/manage/configuration/config.py:392]: `self._write` is no longer there, it has been completely replaced by `_atomic_write`

Comment (ltalirz) [reply aiida/manage/configuration/config.py:392]: ah!

File: tests/manage/configuration/test_config.py (hunk 2 above)
Comment (sphuber) [reply tests/manage/configuration/test_config.py:208]: [l'histoire se répete](https://github.com/aiidateam/aiida-core/pull/1666#pullrequestreview-130247279)
TLDR: blame MacOS :)

Comment (ltalirz) [reply tests/manage/configuration/test_config.py:208]: Not quite sure what type of time resolution @giovannipizzi was referring to, but on my MacOS it is quite substantially better

```ipython
In [1]: from os.path import getmtime

In [5]: from time import sleep

In [7]: from pathlib import Path

In [8]: a = Path('a')

In [9]: for i in range(10):
   ...:     sleep(0.1)
   ...:     a.touch()
   ...:     print(getmtime('a'))
   ...:
1606909606.250035
1606909606.350413
1606909606.452756
1606909606.556789
1606909606.658411
1606909606.758832
1606909606.864015
1606909606.968789
1606909607.069461
1606909607.171968

In [10]: for i in range(10):
    ...:     sleep(0.01)
    ...:     a.touch()
    ...:     print(getmtime('a'))
    ...:
1606909614.754383
1606909614.765453
1606909614.775664
1606909614.785863
1606909614.796015
1606909614.806161
1606909614.817056
1606909614.827201
1606909614.837322
1606909614.848603

In [11]: for i in range(10):
    ...:     sleep(0.001)
    ...:     a.touch()
    ...:     print(getmtime('a'))
    ...:
1606909621.740552
1606909621.741794
1606909621.74325
1606909621.744598
1606909621.745944
1606909621.747289
1606909621.748634
1606909621.749978
1606909621.751066
1606909621.752403
In [12]: for i in range(10):
    ...:     sleep(0.0001)
    ...:     a.touch()
    ...:     print(getmtime('a'))
    ...:
1606909728.212414
1606909728.212609
1606909728.212802
1606909728.213136
1606909728.213311
1606909728.213494
1606909728.21371
1606909728.213929
1606909728.21422
1606909728.214399

```

Comment (sphuber) [reply tests/manage/configuration/test_config.py:208]: If this problem has been addressed in at least some of the MacOS versions, maybe it would be best to remove these sleeps. Note that this is not the only place where I am adding it. It is used for a lot of the `verdi` tests that need to use the interactive writer. I would leave that for a separate PR though. I would be fine to remove it since the only problem it tackles is that certain unit tests will fail on certain MacOS systems. Probably not reason enough to slow down many of these tests all the time.

**Review by ltalirz — APPROVED** [review]
ok, in that case good to merge from my side.

let's follow up on the timing issue though

**Comment by sphuber** [comment]
> ok, in that case good to merge from my side.

Do you think this fixes issue #4604 and so I close it? Or do we leave it open?

**Comment by ltalirz** [comment]
Let's close it and reopen in case it reoccurs

File: tests/manage/configuration/test_config.py (hunk 2 above)
Comment (giovannipizzi) [reply tests/manage/configuration/test_config.py:208]: So: 
- I have fixed the (testing) issue already upstream in click some time ago - this will [appear in v8.0](https://github.com/pallets/click/commit/97fde8bf50058c19a77d4de46b9c54b1731b1750) (it's now in v8.0.0a1 according to GitHub)
- According to my old comments in that commit, the precision is 2 seconds (!!) on FAT32!
- According to my commit, at the time I was using Mac OS X 10.12. In 10.13 Apple switched to [APFS](https://en.wikipedia.org/wiki/Apple_File_System) that I guess fixed this and that must be why Leo does not see it anymore (BTW, it [introduced a much more serious bug IMO](https://github.com/giovannipizzi/concurrent-delete-race-problem) with atomicity of file replacements, that I unveiled when hard-testing disk-objectstore! Still no reply from the developers, though). Since this is only for testing so it matters only to us developers, and I think nowadays probably everybody is developing on at least 10.13, probably we could do a PR and clean up these sleeps.

Comment (giovannipizzi) [reply tests/manage/configuration/test_config.py:208]: Just to bring references: in the official [Microsoft docs](https://docs.microsoft.com/en-us/windows/win32/sysinfo/file-times?redirectedfrom=MSDN) they say

> [...] the resolution of create time on FAT is 10 milliseconds, while write time has a resolution of 2 seconds and access time has a resolution of 1 day, so it is really the access date. The NTFS file system delays updates to the last access time for a file by up to 1 hour after the last access.

And then you guys complain about the 1s resolution on older Macs? :-D

Comment (ltalirz) [reply tests/manage/configuration/test_config.py:208]: Thanks @giovannipizzi for the info, that's good to know!

Does this mean that since we are now always creating a temporary file and renaming it, we actually have a time resolution of 10ms on FAT32 as well?

In general, one could also think of an alternative way of testing this: as long as we know where in our code the config file can be written (e.g. the atomic_write function), we could use `pytest-mock` to mock this function and just check whether it is called. Then we don't need to rely on timing updates at the FS level.
